tests/test_parser.py

```diff
@@ -131,3 +134,121 @@ def test_string_utf8():
     parser.consume_byte(b[:1])
     assert parser.valid_next_bytes() == set([Byte(b[1:])])
     parser.consume_byte(b[1:])
+
+
+class TestRecursiveNullableGrammars:
+    """
+    Computing parse tree of recursive nullable grammars will cause an infinite
+    loop if not handled correctly
+    """
+
+    @pytest.mark.timeout(5)
```

Comment:
Not in love with checking for an infinite loop by introducing a timeout, but it wasn't straight-forward to catch this otherwise. Anyone have a suggestion?

5 seconds *should* be more than ample to compute the parse tree...

Reply:
I spent some time thinking about this, and couldn't come up with anything better 🤷